rllib/algorithms/ddpg/ddpg_tf_policy.py

```diff
@@ -454,61 +88,362 @@ def update_target_fn(tau):
         self.update_target(tau=1.0)
 
     # Support both hard and soft sync.
-    def update_target(self, tau: int = None) -> None:
+    def update_target(
+        self: Union[DynamicTFPolicyV2, EagerTFPolicyV2], tau: int = None
+    ) -> None:
         self._do_update(np.float32(tau or self.config.get("tau")))
 
     @override(TFPolicy)
-    def variables(self) -> List[TensorType]:
+    def variables(self: Union[DynamicTFPolicyV2, EagerTFPolicyV2]) -> List[TensorType]:
         return self.model.variables() + self.target_model.variables()
 
 
-def setup_late_mixins(
-    policy: Policy,
-    obs_space: gym.spaces.Space,
-    action_space: gym.spaces.Space,
-    config: AlgorithmConfigDict,
-) -> None:
-    TargetNetworkMixin.__init__(policy, config)
-
-
-def validate_spaces(
-    policy: Policy,
-    observation_space: gym.spaces.Space,
-    action_space: gym.spaces.Space,
-    config: AlgorithmConfigDict,
-) -> None:
-    if not isinstance(action_space, Box):
-        raise UnsupportedSpaceException(
-            "Action space ({}) of {} is not supported for "
-            "DDPG.".format(action_space, policy)
-        )
-    elif len(action_space.shape) > 1:
-        raise UnsupportedSpaceException(
-            "Action space ({}) of {} has multiple dimensions "
-            "{}. ".format(action_space, policy, action_space.shape)
-            + "Consider reshaping this into a single dimension, "
-            "using a Tuple action space, or the multi-agent API."
-        )
-
-
-DDPGTFPolicy = build_tf_policy(
-    name="DDPGTFPolicy",
-    get_default_config=lambda: ray.rllib.algorithms.ddpg.ddpg.DEFAULT_CONFIG,
-    make_model=build_ddpg_models,
-    action_distribution_fn=get_distribution_inputs_and_class,
-    loss_fn=ddpg_actor_critic_loss,
-    stats_fn=build_ddpg_stats,
-    postprocess_fn=postprocess_nstep_and_prio,
-    compute_gradients_fn=gradients_fn,
-    apply_gradients_fn=build_apply_op,
-    extra_learn_fetches_fn=lambda policy: {"td_error": policy.td_error},
-    validate_spaces=validate_spaces,
-    before_init=setup_early_mixins,
-    before_loss_init=setup_mid_mixins,
-    after_init=setup_late_mixins,
-    mixins=[
-        TargetNetworkMixin,
-        ActorCriticOptimizerMixin,
-        ComputeTDErrorMixin,
-    ],
-)
+# We need this builder function because we want to share the same
+# custom logics between TF1 dynamic and TF2 eager policies.
+def get_ddpg_tf_policy(base: Type[Union[DynamicTFPolicyV2, EagerTFPolicyV2]]) -> Type:
+    """Construct a SimpleQTFPolicy inheriting either dynamic or eager base policies.
```

Comment:
Although linter does not complain: We commonly place a black line between sections of docstrings.

Reply:
SimpleQTFPolicy